app/assets/javascripts/discourse/app/components/d-modal.gjs

```diff
@@ -0,0 +1,357 @@
+    }
+
+    // skip when in a form or a textarea element
+    if (
+      event.target.closest("form") ||
+      document.activeElement?.nodeName === "TEXTAREA"
+    ) {
+      return false;
+    }
+
+    return true;
+  }
+
+  @action
+  handleTouchStart(event) {
+    this.startY = event.touches[0].clientY;
+  }
+
+  @action
+  handleTouchEnd() {
+    if (this.closing) {
+      // if the modal is closing we don't want to risk resetting the position
+      // as the user releases the swipe at the same time
+      return;
+    }
+
+    this.#animateWrapperPosition(0);
+  }
+
+  @action
+  handleTouchMove(event) {
+    const verticalScroll = this.startY - event.touches[0].clientY;
+
+    if (verticalScroll > MIN_SWIPE_THRESHOLD) {
+      this.#animateWrapperPosition(0);
+      return;
+    }
+
+    if (
+      verticalScroll > MAX_SWIPE_THRESHOLD &&
+      verticalScroll <= MIN_SWIPE_THRESHOLD
+    ) {
+      this.#animateWrapperPosition(Math.abs(verticalScroll));
+      return;
+    }
+
+    this.closeModal(CLOSE_INITIATED_BY_SWIPE_DOWN);
+  }
+
+  @action
+  handleWrapperClick(e) {
+    if (e.button !== 0) {
+      return; // Non-default mouse button
+    }
+
+    if (!this.dismissable) {
+      return;
+    }
+
+    return this.closeModal(CLOSE_INITIATED_BY_CLICK_OUTSIDE);
+  }
+
+  @action
+  closeModal(initiatedBy) {
+    if (!this.args.closeModal) {
+      return;
+    }
+
+    this.closing = true;
+
+    this.closeModalHandler = later(() => {
+      this.args.closeModal({ initiatedBy });
+      this.closing = false;
+    }, CLOSE_MODAL_DURATION);
+  }
+
+  @action
+  handleDocumentKeydown(event) {
+    if (this.args.hidden) {
+      return;
+    }
+
+    if (event.key === "Escape" && this.dismissable) {
+      event.stopPropagation();
+      this.closeModal(CLOSE_INITIATED_BY_ESC);
+    }
+
+    if (event.key === "Enter" && this.shouldTriggerClickOnEnter(event)) {
+      this.wrapperElement
+        .querySelector(".d-modal__footer .btn-primary")
+        ?.click();
+      event.preventDefault();
+    }
+  }
+
+  @action
+  handleCloseButton() {
+    this.closeModal(CLOSE_INITIATED_BY_BUTTON);
+  }
+
+  @action
+  validateFlashType(type) {
+    if (type && !FLASH_TYPES.includes(type)) {
+      throw `@flashType must be one of ${FLASH_TYPES.join(", ")}`;
+    }
+  }
+
+  // Could be optimised to remove classic component once RFC389 is implemented
+  // https://rfcs.emberjs.com/id/0389-dynamic-tag-names
+  @cached
+  get dynamicElement() {
+    const tagName = this.args.tagName || "div";
+    if (!["div", "form"].includes(tagName)) {
+      throw `@tagName must be form or div`;
+    }
+
+    return class WrapperComponent extends ClassicComponent {
+      tagName = tagName;
+    };
+  }
+
+  #animateWrapperPosition(position) {
+    requestAnimationFrame(() => {
+      // marginTop is used to not interfere with the paddingTop of the
+      // slide up/down animation
+      this.wrapperElement.style.marginTop = `${position}px`;
```

Comment:
Animating these kinds of attributes is very slow because it requires browsers to recalculate styles. For a smoother UX, we should [animate properties that do not cause reflow/repaint](https://developer.mozilla.org/en-US/docs/Learn/Performance/CSS#choosing_properties_to_animate). So in this case, I think we want a transform.

This is what we do for other swipe-in/out menus like the mobile hamburger/user-menu. Dunno how feasible it is, but it would be great to share the logic for all this swiping/momentum/animation.

Reply:
This is what I do for the css part of it, but this is a very small an specific part of the animation here